src/github.com/matrix-org/dendrite/roomserver/query/query.go

```diff
@@ -330,6 +330,62 @@ func (r *RoomserverQueryAPI) QueryInvitesForUser(
 	return nil
 }
 
+// QueryServerAllowedToSeeEvent implements api.RoomserverQueryAPI
+func (r *RoomserverQueryAPI) QueryServerAllowedToSeeEvent(
+	request *api.QueryServerAllowedToSeeEventRequest,
+	response *api.QueryServerAllowedToSeeEventResponse,
+) error {
+	stateEntries, err := state.LoadStateAtEvent(r.DB, request.EventID)
+	if err != nil {
+		return err
+	}
+
+	stateAtEvent, err := r.loadStateEvents(stateEntries)
+	if err != nil {
+		return err
+	}
+
+	// TODO: Should this be lifted out of here to a more general set of
+	// auth functions?
+
+	isInRoom := false
+	for _, ev := range stateAtEvent {
+		membership, err := ev.Membership()
+		if err != nil {
+			continue
+		}
+
+		if membership != "join" {
+			continue
+		}
+
+		stateKey := ev.StateKey()
+		if stateKey == nil {
+			continue
+		}
+
+		_, domain, err := gomatrixserverlib.SplitID('@', *stateKey)
+		if err != nil {
+			continue
+		}
+
+		if domain == request.ServerName {
+			isInRoom = true
+			break
+		}
+	}
+
+	if isInRoom {
+		response.AllowedToSeeEvent = true
+		return nil
+	}
+
+	// TODO: Check if history visibility is shared and if the server is currently in the room
```

Comment:
FWIW I think synapse always checks if the server is currently in the room.

https://github.com/matrix-org/synapse/blob/master/synapse/handlers/federation.py#L1385-L1394